### ISSUE TITLE
ModuleNotFoundError: No module named 'sklearn.neighbors.kde'

### ISSUE THREAD

**Issue by huzuner** [bcgsc/NanoSim#108]
When I run simulator.py script, I have this error:

Traceback (most recent call last):
  File "simulator.py", line 2281, in <module>
    main()
  File "simulator.py", line 2050, in main
    read_profile(ref_g, number, model_prefix, perfect, args.mode, strandness, dna_type=dna_type, chimeric=chimeric)
  File "simulator.py", line 483, in read_profile
    kde_unaligned = joblib.load(model_prefix + "_unaligned_length.pkl")
  File "/home/uzuner/miniconda3/envs/myenv369/lib/python3.6/site-packages/joblib/numpy_pickle.py", line 585, in load
    obj = _unpickle(fobj, filename, mmap_mode)
  File "/home/uzuner/miniconda3/envs/myenv369/lib/python3.6/site-packages/joblib/numpy_pickle.py", line 504, in _unpickle
    obj = unpickler.load()
  File "/home/uzuner/miniconda3/envs/myenv369/lib/python3.6/pickle.py", line 1050, in load
    dispatch[key[0]](self)
  File "/home/uzuner/miniconda3/envs/myenv369/lib/python3.6/pickle.py", line 1338, in load_global
    klass = self.find_class(module, name)
  File "/home/uzuner/miniconda3/envs/myenv369/lib/python3.6/pickle.py", line 1388, in find_class
    __import__(module, level=0)
ModuleNotFoundError: No module named 'sklearn.neighbors.kde'

And my command is that is run in the /src directory:

 python simulator.py genome -rg /home/uzuner/Documents/ReadSimulation/Snakemake/refs/chrY.fasta -c /home/uzuner/Documents/Software/NanoSim/pre-trained_models/human_NA12878_DNA_FAB49712_albacore/training -o test

I have spent 24 hours to solve this problem and now I know the solution. I opened this issue just to help other people who might also have this annoying problem.

The required **scikit-learn** version should be exactly: **0.20.0** The older or newer versions do not work for this script to run.
My suggestion to the developers is that either you need to make it clear in the READ.ME or make it possible to run with newer versions of scikit-learn.

**Comment by cheny19** [comment]
Hi @huzuner ,

Sorry to hear that you spent so much time on the dependencies, and thank you for your patience. 

`scikit-learn` is a required package as listed in the README, but I don't think it is only compatible with one version 0.20.0. I quickly checked my python environment and it works well with v0.21.3 (Python v3.7.4). I've been updating Python packages regularly, and this `ModuleNotFoundError` has never occurred. I think it's possible that certain version of scikit-learn may not have the kde module or it needs to imported in a different way, but it shouldn't be a common issue.

Cheers,
Chen

**Comment by huzuner** [comment]
Helo @cheny19 ,

Thank you for your answer.

> I think it's possible that certain version of scikit-learn may not have the kde module or it needs to imported in a different way, but it shouldn't be a common issue.

I think so. I have even installed nanosim in a completely different system after cloning it on github. Then I have installed all the dependencies in the usual way with pip install without stating a version. Then I have tried the same command but I kept getting the the same error all the time. It got fixed as soon as I installed scikit-learn 0.20.0 as stated in the requirements.txt file. I think there is a problem but I do not have the sufficiency to actually know that.

Again, thanks a lot for your fast and kind reply.

Best
Hamdiye
